src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSSynchronizationContext.cs

```diff
@@ -0,0 +1,150 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Threading;
+using System.Runtime;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+using System.Collections.Concurrent;
+
+namespace System.Runtime.InteropServices.JavaScript {
+    internal sealed unsafe class JSSynchronizationContext : SynchronizationContext {
+        public readonly Thread MainThread;
+
+        private readonly struct WorkItem {
+            public readonly SendOrPostCallback Callback;
+            public readonly object? Data;
+            public readonly int Id;
+
+            public WorkItem (SendOrPostCallback callback, object? data, int id) {
+                Callback = callback;
+                Data = data;
+                Id = id;
+            }
+        }
+
+        private static JSSynchronizationContext? MainThreadSynchronizationContext;
+        private static readonly ConcurrentDictionary<int, byte> PendingWorkItems = new ();
+        private static int PumpIsPending, NextItemId;
```

Comment:
They can't be volatile, it generates a compiler warning which fails the build. I think all the reads and writes are using Volatile or Interlocked anyway, but I'll double-check.

Reply:
As for using an event or a semaphore - we never wait on PumpIsPending, it just determines whether we queue a new pump. So I'm not sure there's any value in using a heavier primitive like that, but I'm open to doing it. I guess we could TryEnter it to query its state?

Reply:
I'm not even particularly invested in any synchronization primitive, if none of them are a good fit.

I think what I'm really after is to isolate the `int` into a helper struct and wrap the `Interlocked`/`Volatile` operations on it in some higher-level functions.  So that it's easier to treat it as some kind of abstraction when reading the code.